doc/c-interface.md

```diff
@@ -0,0 +1,97 @@
+The request and response channels are handshaked according to the following scheme:
+- The initiator asserts `valid`. The assertion of `valid` must not depend on `ready`. The assertion of ready may depend on `valid`.
+- Once `valid` has been asserted all data must remain stable.
+- The receiver asserts `ready` whenever it is ready to receive the transaction. Asserting `ready` by default is allowed. While `valid` is low, `ready` may be retracted at any time.
+- When both `valid` and `ready` are high the transaction is successful.
+
+### Interface parameters
+The interface is parameterized using the following set of parameters.
+
+| Name               | Type / Range        | Description                                      |
+| ------------------ | ------------------- | ------------------------------------------------ |
+| `NumReq`           | `int` (>=1)         | Number of requesting entities                    |
+| `TotNumRsp`           | `int` (>=1)         | Total Number of responding entities                    |
+| `NumHier`          | `int` (>=1)         | Number of hierarchical interconnect levels       |
+| `DataWidth`        | `int` (32, 64, 128) | ISA Bit-width                                    |
+
+#### Derived Parameters
+The expression `idx(num_idx)` is denotes the index width required to represent up to `num_idx` indices as a binary encoded signal.
+```
+idx(num_idx) = (num_idx > 1) ? $clog2(num_idx) : 1;
+```
+
+| Name               | Value                          | Description                                         |
+| ------------------ | ------------------------------ | --------------------------------------------------- |
+| `HierAddrWidth`    | `idx(NumHier)`                 | Hierarchy level address width                       |
+| `AccAddrWidth`     | `idx(NumRsp)`                  | Accelerator address width            |
+| `AddrWidth`        | `HierAddrWidth + AccAddrWidth` | Overall address width                               |
+| `ExtIdWidth`       | `idx(NumReq) + 1`              | ID Tag width at accelerator-end of the interconnect |
+
+### Request Channel (`q`)
+An offload request comprises the entire 32-bit RISC-V instruction three operands and a request ID tag specifying requesting entity.
+The nature of the offloaded instructions is not of importance to the accelerator interconnect.
+The request channel interface signals are:
+
+| Signal Name   | Range                      | Description                                     |
+| ------------- | -------------------------- | ----------------------------------------------- |
+| `q_addr`      | `AddrWidth-1:AccAddrWidth` | Accelerator hierarchy level.                    |
+|               | `AccAddrWidth-1:0`         | Accelerator address.                            |
+| `q_id`        | `ExtIdWidth-1:1`           | Requester ID                                    |
+|               | `0:0`                      | `1'b0`                                           |
```

Comment:
Okay, I remembered now, when at the end of the document you obtain the `rd` from the instruction word. I think we need some further clarification on the ordering model of the accelerator interface. The `id` field originally had the meaning that requests with a different `id` can be processed in a different order.

At least for the Snitch usecase (and I suspect others as well), it is necessary to be flexible with the `id` ordering so that we can process different instructions also out-of-order. That would mean that the same `id` does not imply any ordering constraints. At this stage we are probably also better off re-naming the `id` field because it will cause confusion.

The other option is that we keep the `id` field and we define that transactions to the same accelerator and the same `id` retain their order (that would be the almost AXI compliant way). I think this needs further discussions and that strikes me as an important topic.

Reply:
So, the `id` field in the current definition really only specifies from which core the request originated. It does not imply any ordering anymore at all, since all necessary ordering is taken care of by the adapter using the scoreboard (i.e. no requests may be issued if they depend on or write to a dirty `rd`.
All instructions that left the adapter may in principle be executed in any arbitrary order.

The AXI compliant way you suggest would be possible, and is probably the most common case. But there is really no good reason for this restriction; if it makes sense for an accelerator to execute instructions in another order it should be able to do so.

I think therefore it is best to rename the signal.